File: backend/app/conductor/kernel/memory_sync.py
import asyncio
import logging
import uuid
from dataclasses import dataclass
from typing import Optional

logger = logging.getLogger(__name__)


@dataclass
class MemorySessionEntry:
    session_id: uuid.UUID
    sandbox_db_id: uuid.UUID
    mount_name: str


class MemoryStoreSubscribers:
    """Tracks which sessions subscribe to each memory store.

    Enables cross-session sync: when one session writes to a shared store,
    other sessions sharing that store can be notified to reload.

    Ported from conductor-kernel/src/memory_sync.rs.
    """

    def __init__(self):
        self._inner: dict[uuid.UUID, list[MemorySessionEntry]] = {}
        self._lock = asyncio.Lock()

    async def register(self, store_id: uuid.UUID, entry: MemorySessionEntry) -> None:
        async with self._lock:
            if store_id not in self._inner:
                self._inner[store_id] = []
            self._inner[store_id].append(entry)

    async def unregister_session(self, session_id: uuid.UUID) -> None:
        async with self._lock:
            for entries in self._inner.values():
                entries[:] = [e for e in entries if e.session_id != session_id]
            self._inner = {k: v for k, v in self._inner.items() if v}

    async def get_peers(
        self, store_id: uuid.UUID, exclude_session: uuid.UUID
    ) -> list[MemorySessionEntry]:
        async with self._lock:
            entries = self._inner.get(store_id, [])
            return [e for e in entries if e.session_id != exclude_session]

    async def get_stores_for_session(self, session_id: uuid.UUID) -> list[uuid.UUID]:
        async with self._lock:
            return [
                store_id
                for store_id, entries in self._inner.items()
                if any(e.session_id == session_id for e in entries)
            ]

    async def notify_peers(
        self,
        store_id: uuid.UUID,
        source_session_id: uuid.UUID,
        change_type: str,
        path: str,
    ) -> int:
        """Notify peer sessions that a memory file changed.

        Returns the number of peers notified. The actual notification is sent
        via the SandboxBridge control input channel.
        """
        peers = await self.get_peers(store_id, source_session_id)
        if not peers:
            return 0

        from app.conductor.lifespan import get_bridge_registry
        import json

        registry = get_bridge_registry()
        if not registry:
            return 0

        notified = 0
        for peer in peers:
            bridge = await registry.get(peer.sandbox_db_id)
            if bridge:
                payload = json.dumps({
                    "type": "memory_sync",
                    "store_id": str(store_id),
                    "mount_name": peer.mount_name,
                    "change_type": change_type,
                    "path": path,
                })
                await bridge.send_control_input(f"__conductor_input_v1__:{payload}")
                notified += 1

        if notified:
            logger.debug(
                "Notified %d peers about %s change in store %s (path=%s)",
                notified, change_type, store_id, path,
            )
        return notified

File: backend/app/joysafeter_orchestrator_rs/src/kernel/memory_sync.rs
use std::collections::HashMap;

use tokio::sync::Mutex;
use tracing::debug;
use uuid::Uuid;

use crate::grpc::proto::{self, orchestrator_message, OrchestratorMessage};
use crate::kernel::ha::BridgeStore;

/// In-memory tracking of which sessions subscribe to which memory stores.
///
/// Mirrors the Python `MemoryStoreSubscribers`. When a memory file changes
/// in one sandbox, this module notifies all peer sandboxes that share the
/// same memory store.
pub struct MemoryStoreSubscribers {
    /// Maps store_id → set of (session_id, sandbox_db_id) tuples
    subscriptions: Mutex<HashMap<String, Vec<MemorySubscription>>>,
}

#[derive(Clone, Debug)]
struct MemorySubscription {
    session_id: Uuid,
    sandbox_db_id: Uuid,
    mount_name: String,
    mount_path: String,
}

impl MemoryStoreSubscribers {
    pub fn new() -> Self {
        Self {
            subscriptions: Mutex::new(HashMap::new()),
        }
    }

    /// Register a session's subscription to a memory store.
    pub async fn register(
        &self,
        store_id: &str,
        session_id: Uuid,
        sandbox_db_id: Uuid,
        mount_name: &str,
        mount_path: &str,
    ) {
        let mut subs = self.subscriptions.lock().await;
        let entry = subs.entry(store_id.to_string()).or_default();

        // Avoid duplicate registrations
        if !entry
            .iter()
            .any(|s| s.session_id == session_id && s.sandbox_db_id == sandbox_db_id)
        {
            entry.push(MemorySubscription {
                session_id,
                sandbox_db_id,
                mount_name: mount_name.to_string(),
                mount_path: mount_path.to_string(),
            });
            debug!(
                store_id = store_id,
                session_id = %session_id,
                "Registered memory store subscription"
            );
        }
    }

    /// Unregister all subscriptions for a session/sandbox pair.
    pub async fn unregister(&self, session_id: Uuid, sandbox_db_id: Uuid) {
        let mut subs = self.subscriptions.lock().await;
        for entries in subs.values_mut() {
            entries.retain(|s| !(s.session_id == session_id && s.sandbox_db_id == sandbox_db_id));
        }
    }

    /// Notify all peers of a memory file change.
    ///
    /// Sends `MemoryFileUpdate` to all sandbox bridges that share the same
    /// store, excluding the sender.
    pub async fn notify_peers(
        &self,
        store_mount_name: &str,
        relative_path: &str,
        content: &[u8],
        operation: &str,
        sender_sandbox_id: Uuid,
        bridge_store: &dyn BridgeStore,
    ) {
        // M3 fix: Collect peers under the lock, then drop it before awaiting
        // gRPC sends. Holding a Mutex across await points can cause deadlocks
        // and blocks other tasks from registering/unregistering subscriptions.
        let peers: Vec<(Uuid, String)> = {
            let subs = self.subscriptions.lock().await;
            let mut result = Vec::new();
            for (_store_id, entries) in subs.iter() {
                for sub in entries {
                    if sub.mount_name == store_mount_name && sub.sandbox_db_id != sender_sandbox_id
                    {
                        result.push((sub.sandbox_db_id, sub.mount_name.clone()));
                    }
                }
            }
            result
            // lock dropped here
        };

        for (peer_sandbox_id, mount_name) in peers {
            if let Some(bridge) = bridge_store.get_by_db_id(peer_sandbox_id) {
                let msg = OrchestratorMessage {
                    payload: Some(orchestrator_message::Payload::MemoryUpdate(
                        proto::MemoryFileUpdate {
                            store_mount_name: mount_name,
                            relative_path: relative_path.to_string(),
                            content: content.to_vec(),
                            operation: operation.to_string(),
                        },
                    )),
                };
                let _ = bridge.send_to_runner(msg).await;
                debug!(
                    peer_sandbox = %peer_sandbox_id,
                    path = relative_path,
                    "Sent memory update to peer"
                );
            }
        }
    }

    /// Get all peer sandbox IDs for a store (excluding the given sandbox).
    pub async fn get_peers(&self, store_id: &str, exclude_sandbox: Uuid) -> Vec<Uuid> {
        let subs = self.subscriptions.lock().await;
        subs.get(store_id)
            .map(|entries| {
                entries
                    .iter()
                    .filter(|s| s.sandbox_db_id != exclude_sandbox)
                    .map(|s| s.sandbox_db_id)
                    .collect()
            })
            .unwrap_or_default()
    }
}
